tonic/src/transport/endpoint.rs

```diff
@@ -252,3 +203,96 @@ impl fmt::Debug for Endpoint {
         f.debug_struct("Endpoint").finish()
     }
 }
+
+/// Configures TLS settings for endpoints.
+#[cfg(feature = "tls")]
+pub struct ClientTlsConfig {
+    provider: TlsProvider,
+    domain: Option<String>,
+    cert: Option<Certificate>,
+    #[cfg(feature = "openssl")]
+    openssl_raw: Option<openssl1::ssl::SslConnector>,
+    #[cfg(feature = "rustls")]
+    rustls_raw: Option<tokio_rustls::rustls::ClientConfig>,
+}
+
+#[cfg(feature = "tls")]
+impl fmt::Debug for ClientTlsConfig {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ClientTlsConfig")
+            .field("provider", &self.provider)
+            .finish()
+    }
+}
+
+#[cfg(feature = "tls")]
+impl ClientTlsConfig {
+    /// Creates a new `ClientTlsConfig` backed by the specified provider. Enable the `openssl` or
+    /// `rustls` features of the `tonic` crate to use OpenSSL or Rustls respectively.
+    pub fn new(provider: TlsProvider) -> Self {
+        ClientTlsConfig {
+            provider,
+            domain: None,
+            cert: None,
+            #[cfg(feature = "openssl")]
+            openssl_raw: None,
+            #[cfg(feature = "rustls")]
+            rustls_raw: None,
+        }
+    }
+
+    /// Sets the domain name against which to verify the server's TLS certificate. If set to `None`
+    /// (the default), the address specified
+    pub fn domain_name(mut self, domain_name: impl Into<Option<String>>) -> Self {
+        self.domain = domain_name.into();
+        self
+    }
+
+    /// Sets the CA Certificate against which to verify the server's TLS certificate.
+    pub fn ca_certificate(mut self, ca_certificate: Certificate) -> Self {
```

Comment:
What do you think about this taking `&mut self` and returning `&mut Self` like the other builders?

Reply:
I've had a look at this, but we have the issue that we want to take the `SslConnector` and `ClientConfig` values from the builder? Happy to look at alternatives though.

Reply:
I believe the configs should be cheaply clonable right? So that means we can take a &self and get a full version.